### ISSUE TITLE
Evaluatable expressions with parameters are no more pre-evaluated

### ISSUE THREAD

**Issue by dbakuntsev** [nhibernate/nhibernate-core#2458]
Tried with the `git clone https://github.com/nhibernate/nhibernate-core.git` branch.

The following mock-up code will not work:

```csharp
Guid guid = Guid.Parse("2D7E6EB3-BD08-4A40-A4E7-5150F7895821");
session.Query<Entity>().Where(x => x.TextField.Contains($"VALUE {guid}"))
```

That is, `NhRelinqQueryParser.PreTransform()` instead of evaluating `String.Format` to its final form `"VALUE 2D7E6EB3-BD08-4A40-A4E7-5150F7895821"` will stop short and will yield `Format("VALUE {0}", Convert("2D7E6EB3-BD08-4A40-A4E7-5150F7895821"))` instead. Naturally, that will not be translated to any meaningful SQL later on and will result in an exception `System.NotSupportedException: System.String Format(System.String, System.Object, System.Object)`.

A trivial workaround exists:

```csharp
Guid guid = Guid.Parse("2D7E6EB3-BD08-4A40-A4E7-5150F7895821");
session.Query<Entity>().Where(x => x.TextField.Contains($"VALUE {guid.ToString()}"))
```

But it occurs to me that this is a bug, as the workaround shouldn't be required. 

Furthermore, it is not obvious what's going on from looking at the final exception (so the workaround may be implemented).

**Comment by bahusoid** [comment]
>That is, NhRelinqQueryParser.PreTransform() instead of evaluating String.Format to its final form "VALUE 2D7E6EB3-BD08-4A40-A4E7-5150F7895821" will stop short and will yield Format("VALUE {0}", Convert("2D7E6EB3-BD08-4A40-A4E7-5150F7895821")) instead.

Maybe I'm wrong but  seems should  be reported to re-linq project.

> A trivial workaround exists...

Just a suggestion. Avoid any possible in-code calculations from LINQ query. So in you case you can move value calculation out of LINQ query:
```C#
Guid guid = Guid.Parse("2D7E6EB3-BD08-4A40-A4E7-5150F7895821");
var value = $"VALUE {guid}";
session.Query<Entity>().Where(x => x.TextField.Contains(value))
```

That would simplify query parsing and might avoid you to encounter many unexpected behaviors.

**Comment by dbakuntsev** [comment]
> Maybe I'm wrong but seems should be reported to re-linq project.

I thought so, too, initially. But `NHibernate.Linq.Visitors.NhEvaluatableExpressionFilter` may be a better candidate for the bugfix.

**Comment by maca88** [comment]
This is a regression from #2335, where `NhEvaluatableExpressionFilter.IsEvaluatableUnary` returns `false` in this case as `guid` is detected as a variable. I will make a fix, but most likely tomorrow.

**Comment by dbakuntsev** [comment]
Maybe related:

```csharp

Guid[] guids = { Guid.Parse("07B430F0-C50F-4D8D-9206-073240B602E2") };
session.Query<Entity>().Where(x => guids.Length == 0 || guids.Contains(x.GuidField));

```

This causes a cryptic `System.NotSupportedException: : ( : p1 )`. Specifically, the heartburn is due to `guids.Length == 0` as without it the expression executes just fine. A possible workaround is to use `!guids.Any()` expression instead of `guids.Length == 0`.

**Comment by maca88** [comment]
> Maybe related

It is related. I made a fix for both queries in #2462.
